File: JavaLibrary/src/test/java/com/flat/simplex/support/TokenChain.java
package com.flat.simplex.support;

import com.flat.simplex.lexer.Key;
import com.flat.simplex.lexer.Lexer;
import com.flat.simplex.lexer.Token;
import org.opentest4j.AssertionFailedError;

public class TokenChain {

    private Token start;
    private Token end;

    public static TokenChain readChain(String content) {
        TokenChain chain = new TokenChain();
        chain.start = new Lexer(content).read();
        Token token = chain.start;
        while (token != null) {
            chain.end = token;
            token = token.getNext();
        }

        return chain;
    }

    public static TokenChain mChain() {
        return new TokenChain();
    }

    public static TokenChain mChain(Key key) {
        return new TokenChain().key(key);
    }

    public static TokenChain mChain(String word) {
        return new TokenChain().word(word);
    }

    public static TokenChain mChain(Double number) {
        return new TokenChain().number(number.toString());
    }

    private void addToken(Token token) {
        if (start == null) {
            start = token;
            end = token;
        } else {
            end.setNext(token);
            end = token;
        }

        token = token.getNext();
        while (token != null) {
            end = token;
            token = token.getNext();
        }
    }

    public Token get() {
        return start;
    }

    public TokenChain child(TokenChain token) {
        end.setChild(token.get());
        if ((end.getKey() == Key.Brace && token.end.getKey() == Key.CBrace) ||
                (end.getKey() == Key.Param && token.end.getKey() == Key.CParam) ||
                (end.getKey() == Key.Index && token.end.getKey() == Key.CIndex))
        end.setLastChild(token.end);
        return this;
    }

    public TokenChain token(Token token) {
        addToken(cloneToken(token));
        return this;
    }

    public TokenChain token(TokenChain token) {
        addToken(cloneToken(token.get()));
        return this;
    }

    private Token cloneToken(Token token) {
        Token clone = new Token(token.getString(), 0, token.getLength(), token.getKey());
        if (token.getChild() != null) {
            clone.setChild(cloneToken(token.getChild()));
            if (token.getLastChild() != null) {
                Token findLast = clone.getChild();
                while (findLast.getNext() != null) {
                    findLast = findLast.getNext();
                }
                clone.setLastChild(findLast);
            }
        }
        Token tClone = clone;
        token = token.getNext();
        while (token != null) {
            Token tClone2 = cloneToken(token);
            tClone.setNext(tClone2);
            tClone = tClone2;
            token = token.getNext();
        }
        return clone;
    }

    public TokenChain key(Key key) {
        addToken(new Token(key.name, 0, key.name.length(), key));
        return this;
    }

    public TokenChain keyword(Key key, String word) {
        addToken(new Token(word, 0, word.length(), key));
        return this;
    }

    public TokenChain word(String word) {
        addToken(new Token(word, 0, word.length(), Key.Word));
        return this;
    }

    public TokenChain number(String number) {
        addToken(new Token(number, 0, number.length(), Key.Number));
        return this;
    }

    public TokenChain string(String string) {
        string = "\"" + string + "\"";
        addToken(new Token(string, 0, string.length(), Key.String));
        return this;
    }

    public static void assertOne(String expected, Token actual, String message) {
        TokenChain chain = readChain(expected);
        assertOne(chain.get(), actual, message);
    }

    public static void assertOne(Token expected, Token actual, String message) {
        if (expected != null && actual != null) {
            if (!expected.equals(actual) || ((expected.getChild() == null) != (actual.getChild() == null))) {
                throw new AssertionFailedError(message, chainString(expected), chainString(actual));
            }
            if (expected.getChild() != null && actual.getChild() != null) {
                assertChildChain(expected, actual, expected.getChild(), actual.getChild(), message);
            }
        }
        if ((expected == null) != (actual == null)) {
            throw new AssertionFailedError(message, chainString(expected), chainString(actual));
        }
    }

    public static void assertChain(String expected, Token actual, Token end, String message) {
        TokenChain chain = readChain(expected);
        assertChain(chain.get(), null, actual, end, message);
    }

    public static void assertChain(Token expected, Token endA, Token actual, Token endB, String message) {
        Token a = expected;
        Token b = actual;
        while (a != endA && b != endB) {
            if (!a.equals(b) || ((a.getChild() == null) != (b.getChild() == null))) {
                throw new AssertionFailedError(message, chainString(expected, endA), chainString(actual, endB));
            }
            if (a.getChild() != null && b.getChild() != null) {
                assertChildChain(expected, actual, a.getChild(), b.getChild(), message);
            }
            a = a.getNext();
            b = b.getNext();
        }
        if ((a == endA) != (b == endB)) {
            throw new AssertionFailedError(message, chainString(expected, endA), chainString(actual, endB));
        }
    }

    public static void assertChain(Token expected, Token actual, String message) {
        assertChain(expected, null, actual, null, message);
    }

    private static void assertChildChain(Token pExpected, Token pActual, Token expected, Token actual, String message) {
        Token a = expected;
        Token b = actual;
        while (a != null && b != null) {
            if (!a.equals(b) || ((a.getChild() == null) != (b.getChild() == null))) {
                throw new AssertionFailedError(message, chainString(pExpected), chainString(pActual));
            }
            if (a.getChild() != null && b.getChild() != null) {
                assertChildChain(pExpected, pActual, a.getChild(), b.getChild(), message);
            }
            a = a.getNext();
            b = b.getNext();
        }
        if ((a == null) != (b == null)) {
            throw new AssertionFailedError(message, chainString(pExpected), chainString(pActual));
        }
    }

    private static String chainString(Token token, Token end) {
        StringBuilder sb = new StringBuilder();
        while (token != null && token != end) {
            sb.append(token);
            if (token.getChild() != null) {
                sb.append(" ").append(chainString(token.getChild()));
            }
            token = token.getNext();
            if (token != null) sb.append(" ");
        }
        return sb.toString();
    }

    private static String chainString(Token token) {
        return chainString(token, null);
    }
}

File: SimplexApp/simplex/lang/support/TokenChain.js
var Key;
var Token;
var Lexer;
var AssertError;

class TokenChain {
    constructor() {
        this.start = null;
        this.end = null;
    }
    
    addToken(token) {
        if (this.start === null) {
            this.start = token;
            this.end = token;
        } else {
            this.end.setNext(token);
            this.end = token;
        }
        token = token.getNext();
        while (token !== null) {
            this.end = token;
            token = token.getNext();
        }
    }

    get() {
        return this.start;
    }

    child(token) {
        this.end.setChild(token.get());
        if ((this.end.getKey() === Key.Brace && token.end.getKey() === Key.CBrace) ||
            (this.end.getKey() === Key.Param && token.end.getKey() === Key.CParam) ||
            (this.end.getKey() === Key.Index && token.end.getKey() === Key.CIndex)) {
            this.end.setLastChild(token.end);
        }
        return this;
    }

    token(token) {
        this.addToken(this.cloneToken(token.get()));
        return this;
    }

    cloneToken(token) {
        let clone = new Token(token.getString(), 0, token.getLength(), token.getKey());

        if (token.getChild() !== null) {
            clone.setChild(this.cloneToken(token.getChild()));
            if (token.getLastChild() !== null) {
                let findLast = clone.getChild();
                while (findLast.getNext() !== null) {
                    findLast = findLast.getNext();
                }
                clone.setLastChild(findLast);
            }
        }

        let tClone = clone;
        token = token.getNext();
        while (token !== null) {
            let tClone2 = this.cloneToken(token);
            tClone.setNext(tClone2);
            tClone = tClone2;
            token = token.getNext();
        }
        return clone;
    }

    key(key) {
        this.addToken(new Token(key.name, 0, key.name.length, key));
        return this;
    }

    keyword(key, word) {
        this.addToken(new Token(word, 0, word.length, key));
        return this;
    }

    word(word) {
        this.addToken(new Token(word, 0, word.length, Key.Word));
        return this;
    }

    number(number) {
        this.addToken(new Token(number, 0, number.length, Key.Number));
        return this;
    }

    string(string) {
        string = '\'' + string + '\'';
        this.addToken(new Token(string, 0, string.length, Key.String));
        return this;
    }

    static readChain(content) {
        let chain = new TokenChain();
        chain.start = new Lexer(content).read();
        let token = chain.start;
        while (token !== null) {
            chain.end = token;
            token = token.getNext();
        }
        return chain;
    }

    static mChain(arg) {
        if (typeof arg === 'string') {
            return new TokenChain().word(arg.toString());
        } else if (typeof arg === 'number') {
            return new TokenChain().number(arg.toString());
        } else if (arg instanceof Key) {
            return new TokenChain().key(arg);
        } else {
            return new TokenChain();
        }
    }

    static assertOne (expected, actual, message) {
        if (typeof expected === 'string') expected = this.readChain(expected).get();

        if (expected !== null && actual !== null) {
            if (!expected.equals(actual) || ((expected.getChild() === null) !== (actual.getChild() === null))) {
                throw new AssertError(message, TokenChain.chainString(expected), TokenChain.chainString(actual));
            }
            if (expected.getChild() !== null && actual.getChild() !== null) {
                this.assertChildChain(expected, actual, expected.getChild(), actual.getChild(), message);
            }
        }
        if ((expected === null) !== (actual === null)) {
            throw new AssertError(message, TokenChain.chainString(expected), TokenChain.chainString(actual));
        }
    };

    static assertChain$3(expected, actual, message) {
        TokenChain.assertChain$5(expected, null, actual, null, message);
    }

    static assertChain$4(expected, actual, end, message) {
        let chain = this.readChain(expected);
        TokenChain.assertChain$5(chain.get(), null, actual, end, message);
    }

    static assertChain$5(expected, endA, actual, endB, message) {
        if (typeof expected === 'string') expected = this.readChain(expected);
        let a = expected;
        let b = actual;
        while (a !== endA && b !== endB) {
            if (!a.equals(b) || ((a.getChild() === null) !== (b.getChild() === null))) {
                throw new AssertError(message, TokenChain.chainString(expected, endA), TokenChain.chainString(actual, endB));
            }
            if (a.getChild() !== null && b.getChild() !== null) {
                this.assertChildChain(expected, actual, a.getChild(), b.getChild(), message);
            }
            a = a.getNext();
            b = b.getNext();
        }
        if ((a === endA) !== (b === endB)) {
            throw new AssertError(message, TokenChain.chainString(expected, endA), TokenChain.chainString(actual, endB));
        }
    }

    static assertChain(...args$) {
        switch (args$.length) {
            case 3:
                return TokenChain.assertChain$3(...args$);
            case 4:
                return TokenChain.assertChain$4(...args$);
            case 5:
                return TokenChain.assertChain$5(...args$);
        }
    }

    static assertChildChain(pExpected, pActual, expected, actual, message) {
        let a = expected;
        let b = actual;
        while (a !== null && b !== null) {
            if (!a.equals(b) || ((a.getChild() === null) !== (b.getChild() === null))) {
                throw new AssertError(message, TokenChain.chainString(pExpected), TokenChain.chainString(pActual));
            }
            if (a.getChild() !== null && b.getChild() !== null) {
                this.assertChildChain(pExpected, pActual, a.getChild(), b.getChild(), message);
            }
            a = a.getNext();
            b = b.getNext();
        }
        if ((a === null) !== (b === null)) {
            throw new AssertError(message, TokenChain.chainString(pExpected), TokenChain.chainString(pActual));
        }
    }

    static chainString$1 (token) {
        return TokenChain.chainString$2(token, null);
    }

    static chainString$2(token, end) {
        let sb = '';
        while (token !== null && token !== end) {
            sb += token.toString();
            if (token.getChild() !== null) {
                sb += ' ' + TokenChain.chainString(token.getChild());
            }
            token = token.getNext();
            if (token !== null) {
                sb += ' ';
            }
        }
        return sb;
    }

    static chainString(...args$) {
        switch (args$.length) {
            case 1:
                return TokenChain.chainString$1(...args$);
            case 2:
                return TokenChain.chainString$2(...args$);
        }
    }
}

module.exports = TokenChain;
Key = require("simplex/lang/lexer/Key.js");
Token = require("simplex/lang/lexer/Token.js");
Lexer = require("simplex/lang/lexer/Lexer.js");
AssertError = require("simplex/lang/support/AssertError.js");
